src/main/java/org/broadinstitute/hellbender/utils/haplotype/HaplotypeBAMWriter.java

```diff
@@ -0,0 +1,184 @@
+    }
+
+    /**
+     * Write out read aligned to haplotype to the BAM file
+     *
+     * @param originalRead the read we want to write aligned to the reference genome
+     */
+    protected void writeReadAgainstHaplotype(final GATKRead originalRead) {
+        output.add(originalRead);
+    }
+
+    /**
+     * Write out haplotypes as reads to the BAM, marking specifically those that are among the best haplotypes
+     *
+     * @param haplotypes a collection of haplotypes to write to the BAM
+     * @param bestHaplotypes a subset of haplotypes that contains those that are best "either good or called"
+     * @param paddedReferenceLoc the genome loc of the padded reference
+     */
+    protected void writeHaplotypesAsReads(final Collection<Haplotype> haplotypes,
+                                          final Set<Haplotype> bestHaplotypes,
+                                          final GenomeLoc paddedReferenceLoc) {
+        if (getWriteHaplotypes()) {
+            for (final Haplotype haplotype : haplotypes) {
+                writeHaplotype(haplotype, paddedReferenceLoc, bestHaplotypes.contains(haplotype));
+            }
+        }
+    }
+
+    /**
+     * Write out a representation of this haplotype as a read
+     *
+     * @param haplotype a haplotype to write out.  Cannot be null
+     * @param paddedRefLoc the reference location.  Cannot be null
+     * @param isAmongBestHaplotypes true if among the best haplotypes, false if it was just one possible but not so good
+     */
+    private void writeHaplotype(final Haplotype haplotype,
+                                final GenomeLoc paddedRefLoc,
+                                final boolean isAmongBestHaplotypes) {
+        final SAMRecord record = new SAMRecord(output.getOutHeader());
+        record.setReadBases(haplotype.getBases());
+        record.setAlignmentStart(paddedRefLoc.getStart() + haplotype.getAlignmentStartHapwrtRef());
+        record.setBaseQualities(Utils.dupBytes((byte) '!', haplotype.getBases().length));
+        record.setCigar(AlignmentUtils.consolidateCigar(haplotype.getCigar()));
+        record.setMappingQuality(isAmongBestHaplotypes ? 60 : 0);
+        record.setReadName("HC" + uniqueNameCounter++);
+        record.setAttribute(HAPLOTYPE_SAMPLE_TAG,haplotype.hashCode());
+        record.setReadUnmappedFlag(false);
+
+        record.setReferenceIndex(paddedRefLoc.getContigIndex());
+        record.setAttribute(SAMTag.RG.toString(), READ_GROUP_ID);
+        //TODO - this was in GATK - why is the reverse strand flag set for these ?
+        //record.setFlags(16);
+        record.setFlags(SAMFlag.READ_REVERSE_STRAND.intValue());
+
+        output.add(new SAMRecordToGATKReadAdapter(record));
+    }
+
+    public boolean getWriteHaplotypes() {
+        return writeHaplotypes;
+    }
+
+    public void setWriteHaplotypes(final boolean writeHaplotypes) {
```

Comment:
eliminate the setter and make the field a constructor parameter (default to true). Make the field final.